src/sctools/barcode.py

```
@@ -0,0 +1,134 @@
+import itertools
+from collections import Counter, Mapping
+import numpy as np
+from .encodings import TwoBit
+from .stats import base4_entropy
+
+
+class BarcodeBase:
+    """Abstract base class for sets of barcodes, encoded in 2bit representation for fast operations
+
+    :method summarize_hamming_distances:  returns descriptive statistics on hamming distances
+      between pairs of barcodes
+
+    """
+
+    def __init__(self, barcode_set, barcode_length):
+        """
+        :param Counter barcode_set: dictionary
+        :param int barcode_length: the length of all barcodes in the set. Different-length barcodes
+          are not supported
+        """
+        if not isinstance(barcode_set, Mapping):
+            raise TypeError('barcode set must be a dict-like object mapping barcodes to counts')
+        self._data = barcode_set
+        self._barcode_length = barcode_length
```

Should we throw an error if barcode_length is None?